internal/moduletest/graph/node_variable.go

```diff
@@ -0,0 +1,160 @@
+	File    *moduletest.File
+}
+
+func (n *NodeVariableDefinition) Name() string {
+	return fmt.Sprintf("var.%s", n.Address)
+}
+
+func (n *NodeVariableDefinition) Referenceable() addrs.Referenceable {
+	return &addrs.InputVariable{Name: n.Address}
+}
+
+func (n *NodeVariableDefinition) Execute(ctx *EvalContext) {
+	if ctx.Stopped() || ctx.Cancelled() {
+		return // don't evaluate anything when stopped or cancelled
+	}
+
+	input, diags := ctx.EvaluateUnparsedVariable(n.Address, n.Config)
+	if input != nil {
+		n.File.AppendDiagnostics(diags)
+		if diags.HasErrors() {
+			ctx.SetVariableStatus(n.Address, moduletest.Error)
+			return
+		}
+	} else {
+		input = &terraform.InputValue{
+			Value: cty.NilVal,
+		}
+	}
+
+	value, moreDiags := terraform.PrepareFinalInputVariableValue(addrs.AbsInputVariableInstance{
+		Module: addrs.RootModuleInstance,
+		Variable: addrs.InputVariable{
+			Name: n.Address,
+		},
+	}, input, n.Config)
+	n.File.AppendDiagnostics(moreDiags)
+	if moreDiags.HasErrors() {
+		ctx.SetVariableStatus(n.Address, moduletest.Error)
+		return
+	}
+
+	ctx.SetVariable(n.Address, &terraform.InputValue{
+		Value:       value,
+		SourceType:  terraform.ValueFromConfig,
+		SourceRange: tfdiags.SourceRangeFromHCL(n.Config.DeclRange),
+	})
+}
+
+var (
+	_ GraphNodeExecutable    = (*NodeVariableExpression)(nil)
+	_ GraphNodeReferencer    = (*NodeVariableExpression)(nil)
+	_ GraphNodeReferenceable = (*NodeVariableExpression)(nil)
+)
+
+type NodeVariableExpression struct {
+	Address string
+	Expr    hcl.Expression
+	File    *moduletest.File
+}
+
+func (n *NodeVariableExpression) Name() string {
+	return fmt.Sprintf("var.%s", n.Address)
+}
+
+func (n *NodeVariableExpression) Referenceable() addrs.Referenceable {
+	return &addrs.InputVariable{Name: n.Address}
+}
+
+func (n *NodeVariableExpression) References() []*addrs.Reference {
+	refs, _ := langrefs.ReferencesInExpr(addrs.ParseRefFromTestingScope, n.Expr)
+	return refs
+}
+
+func (n *NodeVariableExpression) Execute(ctx *EvalContext) {
+	refs, refDiags := langrefs.ReferencesInExpr(addrs.ParseRefFromTestingScope, n.Expr)
+
+	if !ctx.ReferencesCompleted(refs) {
+
+		// before we do anything with the diags, we're going to check that all
+		// of the references we know about were successfully parsed. If not,
+		// then we'll skip actually assessing this node.
+
+		ctx.SetVariableStatus(n.Address, moduletest.Skip)
+		return
+	}
+
+	if ctx.Stopped() || ctx.Cancelled() {
+		return // don't evaluate anything when stopped or cancelled
+	}
+
+	availableVariables := make(map[string]cty.Value)
+	for _, ref := range refs {
+		if input, ok := ref.Subject.(addrs.InputVariable); ok {
+			if variable := ctx.GetVariable(input.Name); variable != nil {
+				availableVariables[input.Name] = variable.Value
+			} else if variable, variableDiags := ctx.EvaluateUnparsedVariable(input.Name, nil); variable != nil {
+				refDiags = refDiags.Append(variableDiags)
+				availableVariables[input.Name] = variable.Value
```

Comment:
If the graph is properly ordered, and perhaps this reference is for a variable definition or expression, then a node for that reference should have been resolved at this point. Why would we need this condition?